File: typescript/packages/core/src/permissions/engine.ts
import type { z } from "zod";
import { assertNever } from "../assert-never";
import type { EffectClass, PermissionMode, PermissionsPolicy } from "../log";
import {
  allMatch,
  anyMatch,
  globMatches,
  isFileTool,
  shellOf,
  workspacePath,
} from "./match";

// the evaluation order. Thread rules come later, with
// permission_rule_added.

type Policy = z.infer<typeof PermissionsPolicy>;
type Mode = z.infer<typeof PermissionMode>;

export type Category = "read_only" | "edit" | "other";

export type PermissionCall = {
  readonly tool: string;
  readonly category: Category;
  readonly input: Readonly<Record<string, unknown>>;
  readonly mode: Mode;
};

export type Decision = {
  readonly decision: "allow" | "deny" | "ask";
  readonly source:
    | "policy"
    | "mode"
    | "protected_path"
    | "self_config_guard"
    | "thread_rule";
  readonly rule?: string;
};

export const DEFAULT_PERMISSIONS: Policy = {
  mode: "default",
  allow: [],
  ask: [],
  deny: [],
  protected_paths: [
    ".git/**",
    ".threads/**",
    ".claude/**",
    ".mcp.json",
    "**/.bashrc",
    "**/.zshrc",
    "**/.profile",
    "**/.gitconfig",
    "**/.ssh/**",
  ],
  allow_bypass: false,
  plan_exit_mode: "default",
};

const EDIT_TOOLS = new Set(["write", "edit", "apply_patch", "notebook_edit"]);
const PLAN_TOOLS = new Set(["ask_user", "exit_plan_mode", "todo_write"]);

/** The permission class of a tool: web and browser tools count as other. */
export function category(
  toolName: string,
  effectClass: z.infer<typeof EffectClass> | undefined,
): Category {
  if (EDIT_TOOLS.has(toolName)) return "edit";
  const web =
    toolName === "web_fetch" ||
    toolName === "web_search" ||
    toolName.startsWith("browser_");
  return effectClass === "read_only" && !web ? "read_only" : "other";
}

/** The first decisive step wins; in dont_ask every ask from steps 4-7 becomes deny. */
export function decide(
  permissions: Policy,
  workspace: string,
  call: PermissionCall,
): Decision {
  if (writesSelfConfig(call))
    return { decision: "deny", source: "self_config_guard" };
  const shell = shellOf(call);
  const denied = anyMatch(permissions.deny, call, workspace, shell);
  if (denied !== undefined)
    return { decision: "deny", source: "policy", rule: denied };
  if (call.mode === "plan" && call.category !== "read_only")
    return {
      decision: PLAN_TOOLS.has(call.tool) ? "allow" : "deny",
      source: "mode",
    };
  const later = laterSteps(permissions, workspace, call);
  return call.mode === "dont_ask" && later.decision === "ask"
    ? { ...later, decision: "deny" }
    : later;
}

// Step 1: config, skills, hooks and schedules load only from the host store,
// which no sandbox path reaches; .threads is where an agent would look for them. A call that
// isn't read_only and names such a path in any argument word is denied: file tools, bash, MCP,
// git and a subagent's task alike. A shell can't be told read from write, so bash is denied too.
// ponytail: a word scan, not a shell evaluator: an obfuscated name (".thr''eads") passes, and
// the host store stays unchanged regardless.
const SELF_CONFIG = /(^|\/)\.threads(\/|$)/;
const WORD_BREAK = /[\s'"`=<>|;&()]+/;

function writesSelfConfig(call: PermissionCall): boolean {
  return (
    call.category !== "read_only" &&
    strings(call.input).some((s) =>
      s.split(WORD_BREAK).some((w) => SELF_CONFIG.test(w)),
    )
  );
}

function strings(value: unknown): readonly string[] {
  if (typeof value === "string") return [value];
  if (Array.isArray(value)) return value.flatMap(strings);
  if (typeof value === "object" && value !== null)
    return Object.values(value).flatMap(strings);
  return [];
}

function laterSteps(
  permissions: Policy,
  workspace: string,
  call: PermissionCall,
): Decision {
  const shell = shellOf(call);
  const path = isFileTool(call.tool)
    ? workspacePath(workspace, call.input)
    : undefined;
  const writes = call.category === "edit" && path !== undefined;
  if (writes && permissions.protected_paths.some((g) => globMatches(g, path)))
    return { decision: "ask", source: "protected_path" };
  const asked = anyMatch(permissions.ask, call, workspace, shell);
  if (asked !== undefined)
    return { decision: "ask", source: "policy", rule: asked };
  const allowed = allMatch(permissions.allow, call, workspace, shell);
  if (allowed !== undefined)
    return { decision: "allow", source: "policy", rule: allowed };
  const inside = !isFileTool(call.tool) || path !== undefined;
  return {
    decision: modeDefault(call.mode, call.category, inside),
    source: "mode",
  };
}

type Column = "read_in" | "read_out" | "edit_in" | "other";

function column(category: Category, inside: boolean): Column {
  if (category === "read_only") return inside ? "read_in" : "read_out";
  return category === "edit" && inside ? "edit_in" : "other";
}

/** Step 7, the mode table. */
function modeDefault(
  mode: Mode,
  category: Category,
  inside: boolean,
): Decision["decision"] {
  const col = column(category, inside);
  if (col === "read_in") return "allow";
  switch (mode) {
    case "plan":
      return col === "read_out" ? "ask" : "deny";
    case "dont_ask":
      return "deny";
    case "default":
      return "ask";
    case "accept_edits":
      return col === "edit_in" ? "allow" : "ask";
    case "bypass":
      return "allow";
    default:
      return assertNever(mode);
  }
}

const RANK = { allow: 0, ask: 1, deny: 2 } as const;

/**
 * A handoff target's decision capped by the principal and host ceiling: the
 * ceiling decides in its own mode, the stricter wins, and a tie reports the target's.
 */
export function decideCapped(
  target: Policy,
  ceiling: Policy,
  workspace: string,
  call: PermissionCall,
): Decision {
  const own = decide(target, workspace, call);
  const capped = decide(ceiling, workspace, { ...call, mode: ceiling.mode });
  return RANK[capped.decision] > RANK[own.decision] ? capped : own;
}

File: python/src/threads/permissions/engine.py
"""The permission fold: the first decisive step wins.

The self-config guard, deny rules, plan mode, protected paths, ask rules, allow rules, then the
mode default. In `dont_ask` every ask from the later steps becomes deny. Hooks and the
principal's authority fold in afterwards, in the loop; nothing here can turn a deny into an
allow.
"""

from collections.abc import Iterable, Mapping, Sequence
from dataclasses import dataclass
from typing import Literal

from pydantic import JsonValue

from threads.log import PermissionMode, PermissionRuleAddedData, Permissions
from threads.permissions import rules, shell
from threads.permissions.rules import Rule, family, parse_rule
from threads.permissions.self_config import touches_config
from threads.result import Ok

type Category = Literal["read_only", "edit", "other"]
type Verdict = Literal["allow", "ask", "deny"]
type Source = Literal[
    "policy", "hook", "default", "self_config_guard", "mode", "protected_path", "thread_rule"
]
"""PermissionDecisionData.source; a test pins it to the generated schema type."""

_PLAN_TOOLS = frozenset({"ask_user", "exit_plan_mode", "todo_write"})
# Columns: read_only inside the workspace, read_only outside, edits inside, everything else.
_MODE_TABLE: Mapping[PermissionMode, tuple[Verdict, Verdict, Verdict, Verdict]] = {
    "plan": ("allow", "ask", "deny", "deny"),
    "dont_ask": ("allow", "deny", "deny", "deny"),
    "default": ("allow", "ask", "ask", "ask"),
    "accept_edits": ("allow", "ask", "allow", "ask"),
    "bypass": ("allow", "allow", "allow", "allow"),
}
_RANK: Mapping[Verdict, int] = {"allow": 0, "ask": 1, "deny": 2}


@dataclass(frozen=True, slots=True)
class Decision:
    decision: Verdict
    source: Source
    rule: str | None = None
    """The matched rule string, when a rule decided."""
    reason: str | None = None
    """A hook's why (the reason of its deny, the rule of its ask), or memory_write's."""


@dataclass(frozen=True, slots=True)
class _Call:
    tool: str
    category: Category
    path: str | None
    """Workspace-relative and normalized; None when absent or outside the workspace."""
    outside: bool
    host: str | None
    agent: str | None
    command: shell.Shell | None
    """Set exactly for bash. A missing command is unparseable."""


@dataclass(frozen=True, slots=True)
class Call:
    """The call being decided: its tool, the category standing in for the tool's class, and
    its arguments (parsed by the tool's input schema before this, but read defensively)."""

    tool: str
    category: Category
    input: JsonValue


def decide(
    permissions: Permissions,
    workspace: str,
    mode: PermissionMode,
    call: Call,
    *,
    thread_rules: Sequence[PermissionRuleAddedData] = (),
) -> Decision:
    if call.category != "read_only" and touches_config(call.input):
        return Decision("deny", "self_config_guard")
    decided = _fold(permissions, mode, _prepare(workspace, call), thread_rules)
    if mode == "dont_ask" and decided.decision == "ask":
        return Decision("deny", decided.source, decided.rule)
    return decided


def decide_capped(
    target: Permissions, ceiling: Permissions, workspace: str, mode: PermissionMode, call: Call
) -> Decision:
    """A handoff target's decision intersected with the principal and host ceiling, decided in
    the ceiling's own mode. The stricter wins; on a tie, the target's."""
    mine = decide(target, workspace, mode, call)
    cap = decide(ceiling, workspace, ceiling.mode, call)
    return cap if _RANK[cap.decision] > _RANK[mine.decision] else mine


def _fold(
    p: Permissions,
    mode: PermissionMode,
    call: _Call,
    thread_rules: Sequence[PermissionRuleAddedData],
) -> Decision:
    denied = _deny_step(p, call, thread_rules)
    if denied is not None:
        return denied
    gated = _gate_step(p, mode, call)
    if gated is not None:
        return gated
    asked = _first_any(_parsed(p.ask), call)
    if asked is not None:
        return Decision("ask", "policy", asked)
    allowed = _allowed(_parsed(p.allow), call)
    if allowed is not None:
        return Decision("allow", "policy", allowed)
    thread_allowed = _allowed(_thread(thread_rules, "allow"), call)
    if thread_allowed is not None:
        return Decision("allow", "thread_rule", thread_allowed)
    return Decision(_MODE_TABLE[mode][_column(call)], "mode")


def _gate_step(p: Permissions, mode: PermissionMode, call: _Call) -> Decision | None:
    """Plan mode, then protected paths: both hold even where an allow rule matches."""
    if mode == "plan" and call.category != "read_only":
        return Decision("allow" if call.tool in _PLAN_TOOLS else "deny", "mode")
    path = call.path
    protected = path is not None and any(rules.glob_matches(g, path) for g in p.protected_paths)
    return Decision("ask", "protected_path") if call.category == "edit" and protected else None


def _deny_step(
    p: Permissions, call: _Call, thread_rules: Sequence[PermissionRuleAddedData]
) -> Decision | None:
    denied = _first_any(_parsed(p.deny), call)
    if denied is not None:
        return Decision("deny", "policy", denied)
    thread_denied = _first_any(_thread(thread_rules, "deny"), call)
    return None if thread_denied is None else Decision("deny", "thread_rule", thread_denied)


def _thread(
    thread_rules: Sequence[PermissionRuleAddedData], decision: Literal["allow", "deny"]
) -> list[Rule]:
    return _parsed(r.rule for r in thread_rules if r.decision == decision)


def _column(call: _Call) -> int:
    if call.category == "read_only":
        return 1 if call.outside else 0
    return 2 if call.category == "edit" and not call.outside else 3


def _parsed(texts: Iterable[str]) -> list[Rule]:
    # Setup rejects invalid rules (permission_rule_invalid); one that slips through never matches.
    return [r.value for r in map(parse_rule, texts) if isinstance(r, Ok)]


def _first_any(candidates: Sequence[Rule], call: _Call) -> str | None:
    """Deny and ask: any simple command matching is enough."""
    return next((r.text for r in candidates if _any_hit(r, call)), None)


def _allowed(candidates: Sequence[Rule], call: _Call) -> str | None:
    """Allow: for bash, only a plain single simple command can match (fail closed: separators,
    escapes, expansions and redirections never do), and never one with a dangerous leading
    assignment. `bash(*)` is the one exception: it matches every command."""
    parsed = call.command
    if parsed is None:
        return next((r.text for r in candidates if _hit(r, call)), None)
    anything = next((r.text for r in candidates if _any_command(r)), None)
    if anything is not None:
        return anything
    if parsed.unparseable or not parsed.plain:
        return None
    first: str | None = None
    for command in parsed.commands:
        match = next((r.text for r in candidates if _bash_hit(r, command)), None)
        if match is None or not command.allowable:
            return None
        first = first or match
    return first


def _any_hit(rule: Rule, call: _Call) -> bool:
    parsed = call.command
    if parsed is None:
        return _hit(rule, call)
    if _any_command(rule) or any(_bash_hit(rule, c) for c in parsed.commands):
        return True
    # Unparseable input: the rule's leading words appearing anywhere in the raw text deny it.
    return parsed.unparseable and rule.names("bash") and _consecutive(rule, parsed.raw_words)


def _any_command(rule: Rule) -> bool:
    """`bash(*)`: every command, parsed or not; the sandbox is the boundary."""
    return rule.names("bash") and rule.specifier == "*"


def _bash_hit(rule: Rule, command: shell.Simple) -> bool:
    if not rule.names("bash"):
        return False
    if rule.specifier is None:
        return True
    words = _rule_words(rule.specifier)
    if rule.specifier.endswith(":*"):
        return command.words[: len(words)] == words
    return command.words == words


def _rule_words(specifier: str) -> tuple[str, ...]:
    commands = shell.parse(specifier.removesuffix(":*")).commands
    return commands[0].words if commands else ()


def _consecutive(rule: Rule, raw: tuple[str, ...]) -> bool:
    if rule.specifier is None:
        return True
    words = _rule_words(rule.specifier)
    n = len(words)
    return n > 0 and any(raw[i : i + n] == words for i in range(len(raw) - n + 1))


def _hit(rule: Rule, call: _Call) -> bool:
    if not rule.names(call.tool):
        return False
    spec = rule.specifier
    if spec is None:
        return True
    kind = family(call.tool)
    if kind == "path":
        return call.path is not None and rules.glob_matches(spec, call.path)
    if kind == "domain":
        return call.host is not None and rules.domain_matches(spec, call.host)
    return kind == "agent" and call.agent == spec


def _prepare(workspace: str, call: Call) -> _Call:
    tool, fields = call.tool, call.input if isinstance(call.input, dict) else {}
    path, url, agent, command = (
        v if isinstance(v := fields.get(k), str) else None
        for k in ("path", "url", "agent", "command")
    )
    rel = None if path is None else rules.relative(workspace, path)
    host = None if url is None else rules.host_of(url)
    parsed = None
    if family(tool) == "bash":
        parsed = shell.parse(command) if command is not None else shell.Shell((), True, False, ())
    return _Call(tool, call.category, rel, path is not None and rel is None, host, agent, parsed)
